packages/wonder-blocks-dropdown/components/dropdown-core-virtualized.js

```diff
@@ -0,0 +1,206 @@
+// @flow
+import * as React from "react";
+import ReactDOM from "react-dom";
+import {VariableSizeList as List} from "react-window";
+
+import DropdownVirtualizedItem from "./dropdown-core-virtualized-item.js";
+import SearchTextInput from "./search-text-input.js";
+import SeparatorItem from "./separator-item.js";
+
+import type {DropdownItem} from "../util/types.js";
+
+import {DROPDOWN_ITEM_HEIGHT, SEARCH_ITEM_HEIGHT} from "../util/constants.js";
+
+type Props = {|
+    /**
+     * The complete list of items that will be virtualized.
+     */
+    data: Array<DropdownItem>,
+
+    /**
+     * The VariableSizeList ref that needs to be passed to access to
+     * react-window's instance methods
+     */
+    listRef?: {
+        current: null | React.ElementRef<typeof List>,
+    },
+|};
+
+type State = {|
+    /**
+     * The list width that needs to be passed in order to let react-window
+     * calculate each item position
+     */
+    width: ?number,
+    height: ?number,
+|};
+
+/**
+ * Maximum visible items inside the dropdown list
+ */
+const MAX_VISIBLE_ITEMS = 10;
+
+/**
+ * A react-window's List wrapper that instantiates the virtualized list and
+ * dynamically calculates the item height depending on the type
+ */
+class DropdownCoreVirtualized extends React.Component<Props, State> {
+    state = {
+        height: this.getHeight(),
+        width: null,
+    };
+
+    componentDidMount() {
+        // Wait for styles to be applied. This way, we can get a more precise
+        // value of the container dimensions.
+        setTimeout(() => {
+            this.setWidth();
+        }, 0);
+    }
+
+    componentDidUpdate(prevProps: Props) {
+        const {data, listRef} = this.props;
+
+        // if the items size has changed, then recalculate each item position
+        if (prevProps.data.length !== data.length) {
+            this.setHeight();
+
+            if (listRef && listRef.current) {
+                // the ref can't associate this instance method
+                // $FlowIgnore
+                listRef.current.resetAfterIndex(1);
+            }
+        }
+    }
+
+    /**
+     * Update container width
+     */
+    setWidth() {
+        const rootNode = ((ReactDOM.findDOMNode(this): any): ?HTMLElement);
+
+        // after the non-virtualized items are rendered, we get the container
+        //  width to pass it to react-window's List
+        if (rootNode) {
+            this.setState({
+                width: rootNode.getBoundingClientRect().width,
```

Comment:
Hmm so, I see that we are sorting the items and get the top 10 longest labels to render but somehow I still see the ellipsis 😭 
The only change I made was:
```

const optionItems = new Array(1000).fill(null).map((_, i) => (<OptionItem
    key={i}
    value={(i + 1).toString()}
    label={`School ${i + 1} in Wizarding World Some more really long labels?`}
/>));
```
![Screen Shot 2019-10-10 at 12 43 23 PM](https://user-images.githubusercontent.com/14991054/66600962-9bcd8700-eb5b-11e9-8e3e-9f83d48ce52d.png)

Reply:
I used your snippet and cannot see the ellipsis. Btw I added the 512 max rule to the width :).